File: AndroidApp/AndroidQuizApp/app/src/main/java/com/example/androidquizapp/QuizViewModel.kt
package com.example.androidquizapp
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import android.util.Log



class QuizViewModel : ViewModel() {

    data class Question(
        val id: Int,
        val text: String,
        val choices: List<Choice>,
        val correctAnswer: String
    ) {
        data class Choice(
            val id: String,
            val text: String
        )
        companion object {
            fun fromFirestoreData(id: Int, data: Map<String, Any>): Question {
                val text = data["text"] as? String ?: ""
                val choicesData = data["choices"] as? List<Map<String, Any>> ?: emptyList()
                val choices = choicesData.map { choiceData ->
                    Choice(
                        id = choiceData["id"] as? String ?: "",
                        text = choiceData["text"] as? String ?: ""
                    )
                }
                val correctAnswer = data["correctAnswer"] as? String ?: ""
                return Question(id, text, choices, correctAnswer)
            }
        }
    }

    private val _currentQuestion = MutableLiveData<Question?>()
    val currentQuestion: LiveData<Question?> = _currentQuestion


    /*
    private val question1 = Question(
        id = 1,
        text = "What is the main programming language used for machine learning?",
        choices = listOf(
            Question.Choice(id = "a", text = "JavaScript"),
            Question.Choice(id = "b", text = "Python"),
            Question.Choice(id = "c", text = "Swift"),
            Question.Choice(id = "d", text = "Kotlin")
        ),
        correctAnswer = "b"
    )

    private val question2 = Question(
        id = 2,
        text = "Which term refers to the process of automatically installing and managing a project's dependencies?",
        choices = listOf(
            Question.Choice(id = "a", text = "dependency injection"),
            Question.Choice(id = "b", text = "package management"),
            Question.Choice(id = "c", text = "build automation"),
            Question.Choice(id = "d", text = "version control")
        ),
        correctAnswer = "b"
    )

    private val question3 = Question(
        id = 3,
        text = "Which term refers to a collection of pre-written code that can be used to perform common tasks?",
        choices = listOf(
            Question.Choice(id = "a", text = "library"),
            Question.Choice(id = "b", text = "variable"),
            Question.Choice(id = "c", text = "function"),
            Question.Choice(id = "d", text = "class")
        ),
        correctAnswer = "a"
    )

    private val question4 = Question(
        id = 4,
        text = "Which programming language uses 'func' for function?",
        choices = listOf(
            Question.Choice(id = "a", text = "JavaScript"),
            Question.Choice(id = "b", text = "Python"),
            Question.Choice(id = "c", text = "Swift"),
            Question.Choice(id = "d", text = "Kotlin")
        ),
        correctAnswer = "c"
    )
    private val questions: List<Question> = listOf(question1, question2, question3, question4)


     */

    private val questionsRepository = QuestionsRepository()
    private val _questions = MutableLiveData<List<Question>>()
    val questions: LiveData<List<Question>> = _questions

    init {
        questionsRepository.questions.observeForever { firestoreQuestions: List<Question> ->
            Log.d("QuizViewModel", "Fetched ${firestoreQuestions.size} questions from Firestore")
            _questions.value = firestoreQuestions
            if (firestoreQuestions.isEmpty()) {
                Log.e("QuizViewModel", "Question list is empty")
            }
            getCurrentQuestion()
        }
    }



    var currentQuestionIndex = 0
    val totalQuestions: Int
        get() = questions.value?.size ?: 0
    var score: Int = 0

    fun getCurrentQuestion(): Question {
        if (questions.value.isNullOrEmpty()) {
            throw IllegalStateException("Question list is empty")
        }
        return questions.value!![currentQuestionIndex]
    }

    fun moveToNextQuestion(): Boolean {
        return if (currentQuestionIndex < totalQuestions - 1) {
            currentQuestionIndex++
            true
        } else {
            false
        }
    }

    fun isAnswerCorrect(selectedAnswer: Int): Boolean {
        val correct = getCurrentQuestion().correctAnswer == getCurrentQuestion().choices[selectedAnswer].id
        if (correct) {
            score++
        }
        return correct
    }
}

File: iOSApp/iOS_QuizApp/iOSQuizApp/QuizViewModel.swift
//
//  QuizViewModel.swift
//  
//
//  Created by Edwin Zhang.
//

import Foundation
import SwiftUI
import Combine

struct Question {
    let id: Int
    let text: String
    let choices: [Choice]
    let correctAnswer: String
    
    static func fromFirestoreData(id: String, data: [String: Any]) -> Question? {
            guard let text = data["text"] as? String,
                  let choicesData = data["choices"] as? [[String: Any]],
                  let correctAnswer = data["correctAnswer"] as? String else {
                return nil
            }
            
            let choices = choicesData.compactMap { choiceData -> Choice? in
                guard let id = choiceData["id"] as? String,
                      let text = choiceData["text"] as? String else {
                    return nil
                }
                return Choice(id: id, text: text)
            }
            
            return Question(id: Int(id) ?? 0, text: text, choices: choices, correctAnswer: correctAnswer)
        }
}

struct Choice: Identifiable {
    let id: String
    let text: String
}

class QuizViewModel: ObservableObject {
    let questionsRepository = QuestionsRepository()
    @Published private(set) var questions: [Question] = []
    @Published private(set) var isLoading: Bool = false
    private var cancellables = Set<AnyCancellable>()
        
    init() {
        questionsRepository.$questions
            .receive(on: DispatchQueue.main)
            .assign(to: \.questions, on: self)
            .store(in: &cancellables)
            
        questionsRepository.$isLoading
            .receive(on: DispatchQueue.main)
            .assign(to: \.isLoading, on: self)
            .store(in: &cancellables)
    }
    
    /*
    @Published private(set) var questions: [Question] = [
        Question(
            id: 1,
            text: "What is the main programming language used for machine learning?",
            choices: [
                Choice(id: "a", text: "JavaScript"),
                Choice(id: "b", text: "Python"),
                Choice(id: "c", text: "Swift"),
                Choice(id: "d", text: "Kotlin")
            ],
            correctAnswer: "b"
        ),
        Question(
            id: 2,
            text: "Which term refers to the process of automatically installing and managing a project's dependencies?",
            choices: [
                Choice(id: "a", text: "dependency injection"),
                Choice(id: "b", text: "package management"),
                Choice(id: "c", text: "build automation"),
                Choice(id: "d", text: "version control")
            ],
            correctAnswer: "b"
        ),
        Question(
            id: 3,
            text: "Which term refers to a collection of pre-written code that can be used to perform common tasks?",
            choices: [
                Choice(id: "a", text: "library"),
                Choice(id: "b", text: "variable"),
                Choice(id: "c", text: "function"),
                Choice(id: "d", text: "class")
            ],
            correctAnswer: "a"
        ),
        Question(
            id: 4,
            text: "Which programming language has 'console.log' command?",
            choices: [
                Choice(id: "a", text: "JavaScript"),
                Choice(id: "b", text: "Kotlin"),
                Choice(id: "c", text: "Swift"),
                Choice(id: "d", text: "Python")
            ],
            correctAnswer: "a"
        ),
        Question(
            id: 5,
            text: "Which term refers to the process of translating source code into machine code?",
            choices: [
                Choice(id: "a", text: "compilation"),
                Choice(id: "b", text: "interpretation"),
                Choice(id: "c", text: "execution"),
                Choice(id: "d", text: "debugging")
            ],
            correctAnswer: "a"
        )
    ]
    */
    
    @Published var isQuizFinished: Bool = false
    @Published var currentQuestionIndex: Int = 0
    private var numberOfCorrectAnswers: Int = 0
    
    var currentQuestion: Question {
            questions[currentQuestionIndex]
    }
        
    var scorePercentage: Int {
            Int(Double(numberOfCorrectAnswers) / Double(questions.count) * 100)
    }
        
    func submitAnswer(id: String) {
        if id == currentQuestion.correctAnswer {
            numberOfCorrectAnswers += 1
        }
        
        if currentQuestionIndex + 1 < questions.count {
            currentQuestionIndex += 1
        } else {
            isQuizFinished = true
        }
    }

    
    func checkIsFinished() {
        if currentQuestionIndex >= questions.count - 1 {
            isQuizFinished = true
        } else {
            isQuizFinished = false
        }
    }
        
    
    func handleSubmit(selectedAnswer: String) {
        submitAnswer(id: selectedAnswer)
        if isQuizFinished {
            // Do nothing, we will navigate to the ScoreView
        } else {
            loadNextQuestion()
        }
    }

    func loadNextQuestion() {
        currentQuestionIndex += 1
    }
    
    func resetQuiz() {
        isQuizFinished = false
        currentQuestionIndex = 0
        numberOfCorrectAnswers = 0
        questionsRepository.fetchQuestions()
    }


}
